### PR TITLE
Add the official implementation of c-TPE

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

> [!TIP]
> Please follow the [Quick TODO list](https://github.com/optuna/optunahub-registry/tree/main?tab=readme-ov-file#quick-todo-list-towards-contribution) to smoothly merge your PR.

## Motivation

This PR is to migrate the official implementation of [c-TPE](https://arxiv.org/abs/2211.14411) into OptunaHub.

## TODO List towards PR Merge

Please remove this section if this PR is not an addition of a new package.
Otherwise, please check the following TODO list:

- [x] Copy `./template/` to create your package
- [x] Replace `<COPYRIGHT HOLDER>` in `LICENSE` of your package with your name
- [x] Fill out `README.md` in your package
- [x] Add import statements of your function or class names to be used in `__init__.py`
- [x] (Optional) Add `from __future__ import annotations` at the head of any Python files that include typing to support older Python versions
- [x] Apply the formatter based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)
- [x] Check whether your module works as intended based on the tips in [`README.md`](https://github.com/optuna/optunahub-registry/tree/main)

### CODE REVIEW

**Comment by nabenabe0928** [comment]
I verified the performance of c-TPE by using the experiment setup of Fig. 3 (Top Row) in [the original paper](https://arxiv.org/pdf/2211.14411).
slide-localization

File: package/samplers/ctpe/README.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    return (c, )
+
+
+sampler = optunahub.load_module(package="samplers/ctpe").cTPESampler(constraints_func=constraints)
+study = optuna.create_study(sampler=sampler)
+study.optimize(objective, n_trials=30)
+print(study.best_trials)
+
+```
+
```
Comment (y0z) [inline package/samplers/ctpe/README.md:70]: It is helpful to describe how to define multiple constraints in cTPE.

File: package/samplers/ctpe/README.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+## Abstract
+
+This package aims to reproduce the TPE algorithm used in the paper published at IJCAI'23:
+
+- [c-TPE: Tree-structured Parzen Estimator with Inequality Constraints for Expensive Hyperparameter Optimization](https://arxiv.org/abs/2211.14411)
+
+The default parameter set of this sampler is the recommended setup from the paper and the experiments in the paper can also be reproduced by this sampler.
+
+## Class or Function Names
```
Comment (y0z) [inline package/samplers/ctpe/README.md:18]: Since [the latest README template](https://github.com/optuna/optunahub-registry/tree/main/template#apis) allows to provide detailed API descriptions, it is better to describe cTPE-specific parameters such as         b_magic_exponent,  min_bandwidth_factor, weight_strategy, etc. if you have time.

**Review by y0z — COMMENTED** [review]
I confirmed the example in README.md works.
I leave some comments.

File: package/samplers/ctpe/README.md (hunk 2 above)
Comment (nabenabe0928) [reply package/samplers/ctpe/README.md:18]: Addressed it!

File: package/samplers/ctpe/README.md (hunk 1 above)
Comment (nabenabe0928) [reply package/samplers/ctpe/README.md:71]: I did not exactly get what you meant, but isn't it possible to simply return two constraint values in `constraints` as this sampler takes the same interface as Optuna, or did you mean I should defer to any Optuna API docs?

**Comment by nabenabe0928** [comment]
<details>
<summary>Verification Code</summary>

```python
from __future__ import annotations

import os
import pickle

import matplotlib.pyplot as plt
import numpy as np
import optuna
import optunahub


plt.rcParams["font.family"] = "Times New Roman"
plt.rcParams["font.size"] = 18
plt.rcParams["mathtext.fontset"] = "stix"  # The setting of math font
plt.rcParams["text.usetex"] = True

# Slice Localization with constraint levels of 0.1, 0.5, and 0.9.
# This information is to reproduce the results of Fig. 3 (Top Row).
n_params_key = "n_params"
runtime_key = "runtime"
benchmark_info = {
    0.1: {runtime_key: 119.38172, n_params_key: 8401.0, "oracle": 0.0010303777},
    0.5: {runtime_key: 366.13934, n_params_key: 52929.0, "oracle": 0.00028238542},
    0.9: {runtime_key: 1113.1345, n_params_key: 229633.0, "oracle": 0.00018452932},
}
data_path = f"{os.environ['HOME']}/hpo_benchmarks/hpolib/slice_localization.pkl"
DATASET = pickle.load(open(data_path, mode="rb"))


class HPOLib:
    def __init__(self, quantile: float, seed: int | None = None) -> None:
        assert quantile in [0.1, 0.5, 0.9]
        self._rng = np.random.RandomState(seed)
        self._thresholds = {
            runtime_key: benchmark_info[quantile][runtime_key],
            n_params_key: benchmark_info[quantile][n_params_key],
        }
        self._oracle = benchmark_info[quantile]["oracle"]

    def reseed(self, seed: int | None = None) -> None:
        self._rng = np.random.RandomState(seed)

    def __call__(self, trial: optuna.Trial) -> float:
        param_indices = [
            trial.suggest_categorical("activation_fn_1", list(range(2))),
            trial.suggest_categorical("activation_fn_2", list(range(2))),
            trial.suggest_int("batch_size", 0, 3),
            trial.suggest_int("dropout_1", 0, 2),
            trial.suggest_int("dropout_2", 0, 2),
            trial.suggest_int("init_lr", 0, 5),
            trial.suggest_categorical("lr_schedule", list(range(2))),
            trial.suggest_int("n_units_1", 0, 5),
            trial.suggest_int("n_units_2", 0, 5),
        ]
        config_id = "".join([str(i) for i in param_indices])
        seed = self._rng.randint(4)
        result = DATASET[config_id]
        loss = result["valid_mse"][seed][100]
        trial.set_user_attr(runtime_key, result[runtime_key][seed])
        trial.set_user_attr(n_params_key, result[n_params_key])
        is_feasible = all(
            trial.user_attrs[key] <= self._thresholds[key] for key in [n_params_key, runtime_key]
        )
        trial.set_user_attr("feasible", is_feasible)
        return loss

    def constraints_func(self, trial: optuna.trial.FrozenTrial) -> tuple[float, float]:
        return [
            trial.user_attrs[key] - self._thresholds[key] for key in [n_params_key, runtime_key]
        ]

    def compute_absolute_percentage_loss(self, study: optuna.Study) -> np.ndarray:
        trials = study.trials
        is_feasible = np.array([t.user_attrs["feasible"] for t in trials])
        loss_vals = np.array([t.value for t in trials])
        loss_vals[~is_feasible] = np.inf
        return (np.minimum.accumulate(loss_vals) - self._oracle) / self._oracle


def collect_results(n_seeds: int, n_trials: int, sampler_name: str) -> dict[float, np.ndarray]:
    if sampler_name == "ctpe":
        package_name = "samplers/ctpe"
        repo_owner = "nabenabe0928"
        mod = optunahub.load_local_module(package=package_name, registry_root="./package/")
        sampler_cls = mod.cTPESampler
    elif sampler_name == "random":
        sampler_cls = lambda constraints_func, seed: optuna.samplers.RandomSampler(seed)
    elif sampler_name == "nsgaii":
        sampler_cls = lambda constraints_func, seed: optuna.samplers.NSGAIISampler(
            seed=seed, constraints_func=constraints_func, population_size=8
        )
    elif sampler_name == "tpe":
        sampler_cls = lambda constraints_func, seed: optuna.samplers.TPESampler(
            multivariate=True, constraints_func=constraints_func, seed=seed
        )
    else:
        assert False, sampler_name

    results = {0.1: [], 0.5: [], 0.9: []}
    for q in results:
        hpolib = HPOLib(quantile=q)
        for seed in range(n_seeds):
            print(f"Start with {sampler_name=}, {q=}, and {seed=}.")
            hpolib.reseed(seed)
            sampler = sampler_cls(constraints_func=hpolib.constraints_func, seed=seed)
            study = optuna.create_study(sampler=sampler)
            study.optimize(hpolib, n_trials=n_trials)
            results[q].append(hpolib.compute_absolute_percentage_loss(study))

    return {k: np.asarray(v) for k, v in results.items()}


def visualize(axes: plt.Axes, n_seeds: int, n_trials: int, sampler_name: str) -> plt.Line2D:
    results = collect_results(n_seeds, n_trials, sampler_name)
    dx = np.arange(n_trials) + 1
    color = {
        "ctpe": "red",
        "nsgaii": "magenta",
        "random": "olive",
        "tpe": "blue",
    }[sampler_name]
    for ax, (q, res) in zip(axes, results.items()):
        m = np.mean(res, axis=0)
        s = np.std(res, axis=0) / np.sqrt(n_seeds)
        line, = ax.plot(dx, m, color=color)
        ax.fill_between(dx, m - s, m + s, color=color, alpha=0.2)
        ax.set_title(f"Quantile: {q:.1f}")

    return line


def main(n_seeds: int, n_trials: int) -> None:
    fig, axes = plt.subplots(
        ncols=3, figsize=(27, 4.5), sharex=True, sharey=True, gridspec_kw={"wspace": 0.05}
    )
    for ax in axes:
        ax.grid(which="minor", color="gray", linestyle=":")
        ax.grid(which="major", color="black")
        ax.set_yscale("log")
        ax.set_xlim(1, n_trials)
        ax.set_ylim(0.1, 2000)

    # sampler_names = ["ctpe", "tpe", "nsgaii", "random"]
    sampler_names = ["ctpe", "tpe", "nsgaii", "random"]
    labels = [
        {"ctpe": "c-TPE", "tpe": "Optuna TPE", "nsgaii": "CNSGA-II", "random": "Random"}[name]
        for name in sampler_names
    ]
    lines = [visualize(axes, n_seeds, n_trials, sampler_name) for sampler_name in sampler_names]
    fig.supxlabel("\# of Trials", y=-0.04)
    fig.supylabel("Absolute Percentage Loss", x=0.09)
    fig.legend(
        handles=lines,
        labels=labels,
        loc="lower center",
        ncol=len(labels),
        fontsize=24,
        bbox_to_anchor=(0.5, -0.25),
    )
    plt.savefig("slice-localization.png", bbox_inches="tight")


optuna.logging.set_verbosity(optuna.logging.CRITICAL)
main(n_seeds=50, n_trials=200)

```

</details>

File: package/samplers/ctpe/README.md (hunk 1 above)
Comment (y0z) [reply package/samplers/ctpe/README.md:76]: Providing an example with multiple constraints like the one below is helpful.
However, adding a link to [the docs](https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-optimize-a-model-with-some-constraints) might be enough.

```python
def constraints(trial: optuna.trial.FrozenTrial) -> tuple[float]:
    x = trial.params["x"]
    y = trial.params["y"]
    c1 = float(np.sin(x) * np.sin(y) + 0.95)
    trial.set_user_attr("c1", c1)
    c2 = ...
    trial.set_user_attr("c2", c2)
    return c1, c2
```

Comment (nabenabe0928) [reply package/samplers/ctpe/README.md:79]: Gotcha, I added the reference!

**Review by y0z — APPROVED** [review]
LGTM

**Comment by y0z** [comment]
Published!
https://hub.optuna.org/samplers/ctpe/
